maven-embedder/src/test/java/org/apache/maven/cli/CLIManagerTest.java

```diff
@@ -41,10 +41,13 @@ public void setup()
     public void spacedOptions()
         throws Exception
     {
-        CommandLine cmdLine = cliManager.parse( "-X -Dx=1 -D y=2 test".split( " " ) );
+        CommandLine cmdLine = cliManager.parse( "-X -Dx=1 -D y=2 -Dz test".split( " " ) );
```

Comment:
This test case is a bit strange.

In the "old" version, this test feeds `["-D", "y=2", "test"]` into the CLI parsing code and expects Commons CLI to "understand" that `y=2` sets a property (in other words, it is a value for the `-D` option), but `test` does not. 

However, `cmdLine.getOptionValues( CLIManager.SET_SYSTEM_PROPERTY )` contains `"test"` as well. It does so because we declared `-D` to allow multiple args. `"y=2"` is the first argument, "test" the second. Commons CLI [cannot tell the difference](https://github.com/apache/commons-cli/blob/master/src/main/java/org/apache/commons/cli/DefaultParser.java#L497) between `"y=2"` being an argument to `-D`, and `"test"` **not** being an argument to `-D`: because we declared `-D` to have multiple arguments, it doesn't know when the arguments stop.

In the "new" version, this does not happen. Why not? Because `-Dz` [is detected as a new option](https://github.com/apache/commons-cli/blob/master/src/main/java/org/apache/commons/cli/DefaultParser.java#L479), with `"z"` as its first value. Commons CLI "closes" reading the previous `-D`. It also [assumes](https://github.com/apache/commons-cli/blob/master/src/main/java/org/apache/commons/cli/DefaultParser.java#L485) that `"z"` is going to be the only argument for the new `-D` so it immediately "closes" the last `-D` as well. `"test"` is then leftover, and interpreted as a program argument.


The difficulty is that we've declared `-D` to have _multiple_ args. That, together with the _value separator_ of `'='`, lets Commons CLI do the key/value splitting. If we would say that `-D` has only _one_ arg for each time that `-D` occurs (which is closer to what we expect), a lot of tests fail. Commons CLI has [a helper method](n) to determine if a token is a Java property declaration, but that method tests for _multiple_ argument values only.

---

The Java executable (`java(1)`) only accepts `"-Dfoo"` and `"-Dfoo=bar"`, but it does not accept `"-D foo=whatever"` (note the space after `-D`) - hence it also doesn't support `-D x=1 y=2`.

If we would express the above statement with Commons CLI, it would be `Option.builder("D").hasArg().valueSeparator('=').build()` **but** that doesn't work because Commons CLI only [treats options with **multiple** arguments as possible Java property style](https://github.com/apache/commons-cli/blob/master/src/main/java/org/apache/commons/cli/DefaultParser.java#L586). I am tempted to consider this last thing a bug in Commons CLI, but I'm not sure about it.

@michael-o, @rfscholte, what are your thoughts on this?

Reply:
My opinion is that we shouldn't try to match a Maven User Property with the Java System Property, they both have their own characteristics. 
You already defined the set of differences, so I would prefer to see this fixed in commons-cli (and it would probably make it possible to keep `--define`)
With Maven the space has always been optional after any flag, I prefer to stay consistent for all.

Reply:
@mthmulders Whatever bug is in CLI, we should not work around it in our code. Report to Commons and have them fix it. What we can do is just to advertise consistent behavior.

Reply:
Thanks for your thoughts on this. I’ve reported [CLI-312](https://issues.apache.org/jira/browse/CLI-312), let's see what happens.

Reply:
When [CLI-312](https://issues.apache.org/jira/browse/CLI-312) is resolved, let's re-assess what we want to do with `--define`.

Reply:
Yeah, but that will take time. Given that there was very little work done on CLI in the last couple of years I don't expect a solution so fast unless we do it ourselves.